host/docs/octoclock_firmware_burner.1
.TH "octoclock_firmware_burner" 1 "3.7.1" UHD "User Commands"
.SH NAME
octoclock_firmware_burner - OctoClock Firmware Burner
.SH DESCRIPTION
Burn firmware images onto an Ettus Research OctoClock device over Ethernet.
.SH SYNOPSIS
.B  octoclock_firmware_burner [OPTIONS]
.SH OPTIONS
This program works best when only an IP address is specified.
.IP "Device IP Address:"
--addr=\fI"Address"\fR
.IP "This help information:"
--help
.IP "Custom Firmware Filepath:"
--fw-path=\fI"filepath"\fR
.IP "List all OctoClock devices without burning"
--list
.SH EXAMPLES
.SS Selecting a custom firmware path
.sp
octoclock_firmware_burner --addr=192.168.10.3 --fw-path=~/custom_octoclock_image.bin
.ft
.fi
.SH SEE ALSO
UHD documentation:
.B http://files.ettus.com/manual/
.LP
Other UHD programs:
.sp
uhd_images_downloader(1) usrp2_card_burner(1) usrp_n2xx_simple_net_burner(1) usrp_x3xx_fpga_burner(1)
.SH AUTHOR
This manual page was written by Nicholas Corgan
for the Debian project (but may be used by others).
.SH COPYRIGHT
Copyright (c) 2014 Ettus Research LLC
.LP
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
.LP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.


CHANGELOG
Change Log for Releases
==============================

## 003.007.002
* Properly flushing PCIe chain on device to prevent stale data.
* Adjusted bus clock rate in FPGA to improve timing.
* Fixed issue where FPGA would fail to load FPGA image over PCIe.
* Fixed incompatibility issue with USRP2 FPGA burner utility.
* Fixed issue where ZPU would report empty NMEA strings from GPSDO.
* Updated some functions to no longer use deprecated Boost calls.
* Fixed issue where libusb would declare 'connection refused'.
* Fixed Windows library package naming.
* Updated documentation.
* Fixed reversed TX / RX colors in X300 LED indicators.
* DBSRX2 now works properly with X300 / X310.
* Updated include files to build on older distros of Linux.
* Fixed issue with 'item32' type converter from GNU Radio.
* Fixed issue where channels on B210 would occasionally not be phase-aligned.
* Fixed problem causing channels on B210 to swap between runs.
* Fixed issue in N-Series devices causing scaling error @ 50 MSps.
* The B200 / B210 now generates an internal PPS if none is provided.
* Improved performance on PCIe for X300.
* B200 / B210 properly reports clock rate (issue with float comparison).
* Fixed issue with filter selection in B200 that caused loss of RX power.
* Removed ref lock check from X300 where it wasn't necessary.

## 003.007.001
* Fixed issue with TVRX2 divider calculation.
* Fixed issue using calibration utilities on B-Side daughterboard in an X3xx.
* Replaced unsafe `sscanf` call in utilities.
* Properly initializing N-Series clock, fixing short transient on device boot-up.
* Improved `--help` output for a number of utilities & examples.
* Improved READMEs for directories.
* Fixed X3xx documentation with some clarifications.
* UHD will now tell you if you have requested a sample rate higher than the transport can deliver.
* Removed work-around necessary in RIO kernel module for zero-copy in PCIe for X3xx devices.
* Fixed issue where X3xx devices would lock-up when on networks with lots of traffic.
* The B2xx FX3 and AD9361 source code is now in UHD.
* Numerous B2xx stability improvements.
* Fixed includes for older OSes (e.g., Fedora 14).
* Fixed includes for older versions of Boost.
* Fixed PPS detection in X3xx with multiple time sources.
* Fixed overflow reporting for X3xx utilities.
* Fixed MTU / frame size detection for X3xx devices.
* Fixed B2xx filter chain causing wrong sample rate in some circumstances.

## 003.007.000
* Introduced USRP X300 and X310 support!
* Releasing a CHDR Dissector for Wireshark analysis
* Improved USRP B200 and B210 stability
* Introducing Integer-N tuning for WBX, SBX, CBX daughterboards
* Introducing support for 120 MHz versions of WBX, SBX, CBX
* Lots of new documentation
* New GPIO example for USRP X300
* Fixed threading bug in USRP B2xx code causing GQRX issue
* General UHD bug fixes & improvements

## 003.006.002
* Fixed bug in timed commands example.
* Improved compatibility for older versions of Boost.
* Fixed segmentation fault issue in converter code.
* Fixed binary-stripping for ARM builds.
* B2xx: Further improvements to code and images.
* Documentation links now installed to Start Menu in Windows.
* B2xx: More robust error checking in host code.
* B2xx: Improved error reporting.
* B2xx: Increase FPGA loading speed on USB 3
* B2xx: Re-programming FPGA without cycle now supported.
* B2xx: Larger RX SPP ceiling to support 16K transfers with larger RX FIFO in new FPGA image
* Updated internal READMEs and documentation.
* N-Series: netburner now accepts relative paths with '~'
* Completely re-written UHD Images Downloader, with numerous bug fixes and new features.

## 003.006.001
* B2xx: Fixed critical bug in 003.006.000 regarding USRP B2xx operation

## 003.006.000
* Many small fixes for bugs revealed by static analysis.
* Introduced support for the USRP B200 / B210

## 003.005.003

* E110:
** Fix FPGA Makefile build typo
* UHD:
** Fixed timespec irrational rate rounding
** Multichan streamer CPU utilization

## 003.005.002

* B100:
** Fix get send buffer timeout
* E1x0:
** Changes to bus timings for S issue
* USRP1:
** Restore broken EEPROM writing


## 003.005.001

* Misc:
** Fixes to images downloader
** Fixes to C++ net burner
** Added sleep(1) to query_gpsdo_sensors
* OSX:
** Fix for socket send code newer OS versions
* Changes from 3.4.5

## 003.005.000

* B100:
** Added timed commands feature
** Incremented FPGA compat number to 11.1
** Incremented firmware compat number to 3
* E1x0:
** Added timed commands feature
** Incremented FPGA compat number to 11.1
* USRP2/N2x0:
** Alternative stream destination on TX
** Incremented FPGA compat number to 10
* N2x0:
** Implemented timed-commands feature
** Implemented fast-commands feature
* SBX/WBX
** Tune with phase sync using timed-commands
* RFX series
** Added calibration utilities support
* General:
** SSE2 conversions for sc8 RX samples
** Added multi-threading to packet converters
** Added automatic images fetcher application

## 003.004.005

* XCVR2450:
** Added XCVR2450 rev2 support
* WBX:
** Added WBX rev4 support
* Misc:
** Fix to rx_samples_to_udp byte count
** stream command enums easy on SWIG

## 003.004.004

* Gen2
** Fix RX and TX DSP scalar adjustments
* B100/E1x0
** Fixed RX ADC IQ inversion swap
** Incremented FPGA compat number to 9.4

## 003.004.003

* USRP2/N2x0:
** Net burner parser fix (windows)
* XCVR2450:
** Fix to disable automatic LO offset on TX
* N2x0:
** Deal with misc exceptions in net burner
* E1x0:
** Changes to add reliability to bus state machine
* USRP1:
** Shutoff the DAC on transmit EOB flags
** Revert 1st nyquist zone DAC calculation

## 003.004.002

* USRP2/N2x0:
** Card and net burner language fixes
** Net burner python v3 code fix
** Net burner IPv6 interface fix
* E1x0:
** Fix for FPGA timing issue with GPMC input
** Incremented FPGA compat number to 9.2
* B100:
** Fix USB wrapper/buffer release race condition
* USRP1:

host/utils/octoclock_firmware_burner.cpp
//
// Copyright 2014 Ettus Research LLC
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

#include <algorithm>
#include <csignal>
#include <iostream>
#include <fstream>
#include <stdexcept>
#include <vector>

#include <boost/foreach.hpp>
#include <boost/asio.hpp>
#include <boost/program_options.hpp>
#include <boost/assign.hpp>
#include <boost/cstdint.hpp>
#include <boost/assign/list_of.hpp>
#include <boost/format.hpp>
#include <boost/filesystem.hpp>
#include <boost/thread.hpp>

#include <uhd/device.hpp>
#include <uhd/transport/udp_simple.hpp>
#include <uhd/types/device_addr.hpp>
#include <uhd/types/time_spec.hpp>
#include <uhd/utils/byteswap.hpp>
#include <uhd/utils/images.hpp>
#include <uhd/utils/paths.hpp>
#include <uhd/utils/safe_main.hpp>

#include "../lib/usrp_clock/octoclock/common.h"
#include "ihexcvt.hpp"

#define MAX_FIRMWARE_SIZE 1024*120
#define BLOCK_SIZE 256
#define UDP_TIMEOUT 5

namespace fs = boost::filesystem;
namespace po = boost::program_options;

using namespace uhd;
using namespace uhd::transport;

static int num_ctrl_c = 0;
void sig_int_handler(int){
    num_ctrl_c++;
    if(num_ctrl_c == 1){
        std::cout << std::endl << "Are you sure you want to abort the image burning? If you do, your "
                                  "OctoClock device will be bricked!" << std::endl
                               << "Press Ctrl+C again to abort the image burning procedure." << std::endl << std::endl;
    }
    else{
        std::cout << std::endl << "Aborting. Your OctoClock device will be bricked." << std::endl
                               << "Refer to http://files.ettus.com/manual/page_octoclock.html#bootloader" << std::endl
                               << "for details on restoring your device." << std::endl;
        exit(EXIT_FAILURE);
    }
}

boost::uint8_t firmware_image[MAX_FIRMWARE_SIZE];
size_t firmware_size = 0;
boost::uint8_t octoclock_data[udp_simple::mtu];
octoclock_packet_t *pkt_in = reinterpret_cast<octoclock_packet_t *>(octoclock_data);
std::string firmware_path;
size_t num_blocks = 0;

/*
 * Functions
 */
void list_octoclocks(){
    device_addrs_t found_octoclocks = device::find(uhd::device_addr_t(), device::CLOCK);

    std::cout << "Available OctoClock devices:" << std::endl;
    BOOST_FOREACH(const device_addr_t &oc, found_octoclocks){
        std::cout << " * " << oc["addr"] << std::endl;
    }
}

/*
 * Manually find bootloader. This sends multiple packets in order to increase chances of getting
 * bootloader before it switches to the application.
 */
device_addrs_t bootloader_find(std::string ip_addr){
    udp_simple::sptr udp_transport = udp_simple::make_connected(ip_addr, BOOST_STRINGIZE(OCTOCLOCK_UDP_CTRL_PORT));

    octoclock_packet_t pkt_out;
    pkt_out.sequence = uhd::htonx<boost::uint32_t>(std::rand());
    pkt_out.code = OCTOCLOCK_QUERY_CMD;
    pkt_out.len = 0;
    size_t len = 0;

    device_addrs_t addrs;

    boost::system_time comm_timeout = boost::get_system_time() + boost::posix_time::milliseconds(3000);
    
    while(boost::get_system_time() < comm_timeout){
        UHD_OCTOCLOCK_SEND_AND_RECV(udp_transport, OCTOCLOCK_QUERY_CMD, pkt_out, len, octoclock_data);
        if(UHD_OCTOCLOCK_PACKET_MATCHES(OCTOCLOCK_QUERY_ACK, pkt_out, pkt_in, len) and
           pkt_in->proto_ver == OCTOCLOCK_BOOTLOADER_PROTO_VER){
            addrs.push_back(device_addr_t());
            addrs[0]["type"] = "octoclock-bootloader";
            addrs[0]["addr"] = udp_transport->get_recv_addr();
            break;
        }
    }

    return addrs;
}

void read_firmware(){
    std::ifstream firmware_file(firmware_path.c_str(), std::ios::binary);
    firmware_file.seekg(0, std::ios::end);
    firmware_size = firmware_file.tellg();
    if(firmware_size > MAX_FIRMWARE_SIZE){
        firmware_file.close();
        throw uhd::runtime_error(str(boost::format("Firmware file too large: %d > %d")
                                     % firmware_size % (MAX_FIRMWARE_SIZE)));
    }
    firmware_file.seekg(0, std::ios::beg);
    firmware_file.read((char*)firmware_image, firmware_size);
    firmware_file.close();

    num_blocks = (firmware_size % BLOCK_SIZE) ? (firmware_size / BLOCK_SIZE)
                                              : ((firmware_size / BLOCK_SIZE) + 1);
}

void burn_firmware(udp_simple::sptr udp_transport){
    octoclock_packet_t pkt_out;
    pkt_out.sequence = uhd::htonx<boost::uint32_t>(std::rand());
    pkt_out.len = uhd::htonx<boost::uint16_t>((boost::uint16_t)firmware_size);
    size_t len = 0, current_pos = 0;

    //Tell OctoClock not to jump to application, wait for us instead
    std::cout << "Telling OctoClock to prepare for firmware download..." << std::flush;
    UHD_OCTOCLOCK_SEND_AND_RECV(udp_transport, PREPARE_FW_BURN_CMD, pkt_out, len, octoclock_data);
    if(UHD_OCTOCLOCK_PACKET_MATCHES(FW_BURN_READY_ACK, pkt_out, pkt_in, len)) std::cout << "ready." << std::endl;
    else{
        std::cout << std::endl;
        throw uhd::runtime_error("Could not get OctoClock in valid state for firmware download.");
    }

    std::cout << std::endl << "Burning firmware." << std::endl;
    pkt_out.code = FILE_TRANSFER_CMD;

    //Actual burning below
    for(size_t i = 0; i < num_blocks; i++){
        pkt_out.sequence++;
        pkt_out.addr = i*BLOCK_SIZE;
        std::cout << "\r * Progress: " << int(double(i)/double(num_blocks)*100)
                  << "% (" << (i+1) << "/" << num_blocks << " blocks)" << std::flush;

        memset(pkt_out.data, 0, BLOCK_SIZE);
        memcpy((void*)(pkt_out.data), &firmware_image[i*BLOCK_SIZE], std::min(int(firmware_size-current_pos), BLOCK_SIZE));
        UHD_OCTOCLOCK_SEND_AND_RECV(udp_transport, FILE_TRANSFER_CMD, pkt_out, len, octoclock_data);
        if(not (UHD_OCTOCLOCK_PACKET_MATCHES(FILE_TRANSFER_ACK, pkt_out, pkt_in, len))){
            std::cout << std::endl;
            throw uhd::runtime_error("Failed to burn firmware to OctoClock!");
        }

        current_pos += BLOCK_SIZE;
    }

    std::cout << "\r * Progress: 100% (" << num_blocks << "/" << num_blocks << " blocks)" << std::endl;
}

void verify_firmware(udp_simple::sptr udp_transport){
    octoclock_packet_t pkt_out;
    pkt_out.proto_ver = OCTOCLOCK_FW_COMPAT_NUM;
    pkt_out.sequence = uhd::htonx<boost::uint32_t>(std::rand());
    size_t len = 0, current_pos = 0;

    std::cout << "Verifying firmware." << std::endl;

    for(size_t i = 0; i < num_blocks; i++){
        pkt_out.sequence++;
        pkt_out.addr = i*BLOCK_SIZE;
        std::cout << "\r * Progress: " << int(double(i)/double(num_blocks)*100)
                  << "% (" << (i+1) << "/" << num_blocks << " blocks)" << std::flush;

        UHD_OCTOCLOCK_SEND_AND_RECV(udp_transport, READ_FW_CMD, pkt_out, len, octoclock_data);
        if(UHD_OCTOCLOCK_PACKET_MATCHES(READ_FW_ACK, pkt_out, pkt_in, len)){
            if(memcmp((void*)(pkt_in->data), &firmware_image[i*BLOCK_SIZE],
                      std::min(int(firmware_size-current_pos), BLOCK_SIZE))){
                std::cout << std::endl;
                throw uhd::runtime_error("Failed to verify OctoClock firmware!");
            }
        }
        else{
            std::cout << std::endl;
            throw uhd::runtime_error("Failed to verify OctoClock firmware!");
        }
    }

    std::cout << "\r * Progress: 100% (" << num_blocks << "/" << num_blocks << " blocks)" << std::endl;
}

void reset_octoclock(const std::string &ip_addr){
    udp_simple::sptr udp_transport = udp_simple::make_connected(ip_addr, BOOST_STRINGIZE(OCTOCLOCK_UDP_CTRL_PORT));

    octoclock_packet_t pkt_out;
    pkt_out.sequence = uhd::htonx<boost::uint32_t>(std::rand());
    size_t len;

    UHD_OCTOCLOCK_SEND_AND_RECV(udp_transport, RESET_CMD, pkt_out, len, octoclock_data);
    if(UHD_OCTOCLOCK_PACKET_MATCHES(RESET_ACK, pkt_out, pkt_in, len)) std::cout << "done." << std::endl;
    else throw uhd::runtime_error("Failed to place device in state to receive firmware.");
}

void finalize(udp_simple::sptr udp_transport){
    octoclock_packet_t pkt_out;
    pkt_out.len = 0;
    pkt_out.sequence = uhd::htonx<boost::uint32_t>(std::rand());
    size_t len = 0;

    std::cout << std::endl << "Telling OctoClock to load application..." << std::flush;
    UHD_OCTOCLOCK_SEND_AND_RECV(udp_transport, FINALIZE_BURNING_CMD, pkt_out, len, octoclock_data);
    if(UHD_OCTOCLOCK_PACKET_MATCHES(FINALIZE_BURNING_ACK, pkt_out, pkt_in, len)) std::cout << "done." << std::endl;
    else std::cout << "no ACK. Device may not have loaded application." << std::endl;
}

int UHD_SAFE_MAIN(int argc, char *argv[]){
    std::string ip_addr;
    po::options_description desc("Allowed options");
    desc.add_options()
        ("help", "Display this help message.")
        ("addr", po::value<std::string>(&ip_addr), "Specify an IP address.")
        ("fw-path", po::value<std::string>(&firmware_path), "Specify a custom firmware path.")
        ("list", "List all available OctoClock devices.")
    ;
    po::variables_map vm;
    po::store(po::parse_command_line(argc, argv, desc), vm);
    po::notify(vm);

    //Print help message
    if(vm.count("help")){
        std::cout << "OctoClock Firmware Burner" << std::endl << std::endl;

        std::cout << "Burns a firmware image file onto an OctoClock device. Specify" << std::endl
                  << "the address of the OctoClock with the --addr option. To burn" << std::endl
                  << "a custom firmware image, use the --fw-path option. Otherwise, the" << std::endl
                  << "utility will use the default image. To list all available" << std::endl
                  << "OctoClock devices without burning firmware, use the --list" << std::endl
                  << "option." << std::endl << std::endl;

        std::cout << desc << std::endl;
        return EXIT_SUCCESS;
    }

    //List all available devices
    if(vm.count("list")){
        list_octoclocks();
        return EXIT_SUCCESS;
    }

    if(not (vm.count("addr"))){
        throw uhd::runtime_error("You must specify an address with the --addr option!");
    }
    udp_simple::sptr udp_transport = udp_simple::make_connected(ip_addr, BOOST_STRINGIZE(OCTOCLOCK_UDP_FW_PORT));

    //If custom path given, make sure it exists
    if(vm.count("fw-path")){
        //Expand tilde usage if applicable
        #ifndef UHD_PLATFORM_WIN32
        if(firmware_path.find("~/") == 0) firmware_path.replace(0,1,getenv("HOME"));
        #endif

        if(not fs::exists(firmware_path)){
            throw uhd::runtime_error(str(boost::format("This filepath does not exist: %s") % firmware_path));
        }
    }
    else firmware_path = find_image_path("octoclock_r4_fw.bin");

    //If Intel hex file detected, convert to binary
    std::string ext = fs::extension(firmware_path);
    if(ext == ".hex"){
        std::cout << "Found firmware at path: " << firmware_path << std::endl;

        //Write firmware .bin file to temporary directory
        fs::path temp_bin = fs::path(fs::path(get_tmp_path()) / str(boost::format("octoclock_fw_%d.bin")
                                                          % time_spec_t::get_system_time().get_full_secs()));
        Hex2Bin(firmware_path.c_str(), temp_bin.string().c_str(), false);

        firmware_path = temp_bin.string();
    }
    else if(ext == ".bin"){
        std::cout << "Found firmware at path: " << firmware_path << std::endl;
    }
    else throw uhd::runtime_error("The firmware file has in improper extension (must be .hex or .bin).");

    std::cout << std::endl << boost::format("Searching for OctoClock with IP address %s...") % ip_addr << std::flush;
    device_addrs_t octoclocks = device::find(str(boost::format("addr=%s") % ip_addr), device::CLOCK);
    if(octoclocks.size() == 1){
        //If in application, quietly reset into bootloader and try to find again
        if(octoclocks[0]["type"] == "octoclock"){
            reset_octoclock(ip_addr);
            boost::this_thread::sleep(boost::posix_time::milliseconds(2000));
            octoclocks = bootloader_find(ip_addr);
            if(octoclocks.size() == 1) std::cout << "found." << std::endl;
            else{
                std::cout << std::endl;
                throw uhd::runtime_error("Could not find OctoClock with given IP address!");
            }
        }
        else std::cout << "found." << std::endl;
    }
    else{
        std::cout << std::endl;
        throw uhd::runtime_error("Could not find OctoClock with given IP address!");
    }

    read_firmware();

    std::signal(SIGINT, &sig_int_handler);

    burn_firmware(udp_transport);
    verify_firmware(udp_transport);
    finalize(udp_transport);

    std::cout << "Waiting for OctoClock to reinitialize..." << std::flush;
    boost::this_thread::sleep(boost::posix_time::milliseconds(2000));
    octoclocks = device::find(str(boost::format("addr=%s") % ip_addr), device::CLOCK);
    if(octoclocks.size() == 1){
        if(octoclocks[0]["type"] == "octoclock-bootloader"){
            std::cout << std::endl;
            throw uhd::runtime_error("OctoClock failed to leave bootloader state.");
        }
        else{
            std::cout << "found." << std::endl << std::endl
                      << "Successfully burned firmware." << std::endl << std::endl;
        }
    }
    else{
        std::cout << std::endl;
        throw uhd::runtime_error("Failed to reinitialize OctoClock.");
    }

    return EXIT_SUCCESS;
}
